In [1]:
import numpy as np
import pandas as pd

**Question 1**
- **What's the version of NumPy that you installed?**

In [2]:
np.__version__

'1.23.0'

**Answer:**

- **The version of numpy installed is 1.23.0**

In [3]:
def get_data_from_url(url):
    data_format_from_url = url.split('.')[-1]
    if data_format_from_url == 'csv':
        df = pd.read_csv(url)
    elif data_format_from_url in ['xls','xlsx']:
        df = pd.read_excel(url)
    elif data_format_from_url == 'json':
        df = pd.read_json(url)
    else:
        print('Data file not supported')
    return df

In [4]:
url = 'https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv'
#url = 'MLzoomcamp_homewwork1_data.csv'

In [5]:
car_price_df = get_data_from_url(url)

**Question 2**
- **How many records are in the dataset?**

In [6]:
car_price_df.shape[0]

11914

**Answer:**

- **There are 11914 records in the data set**

**Question 3**

- **Who are the most frequent car manufacturers (top-3) according to the dataset?**

In [7]:
car_price_df['Make'].value_counts().sort_values(ascending = False )[0:3]

Chevrolet     1123
Ford           881
Volkswagen     809
Name: Make, dtype: int64

**Answer:**

- **Chevrolet, Ford, Volkswagen are the top three most frequent car manufacturers**

**Question 4**
- **What's the number of unique Audi car models in the dataset?**

In [8]:
car_price_df[car_price_df['Make']=='Audi']['Model'].nunique()

34

**Answer:**

- **There are 34 unique Audi car models in the dataset**

**Question 5**

- **How many columns in the dataset have missing values?**

In [9]:
car_price_df.isnull().any().sum()

5


**Answer:**

- **There 5 columns in the dataset having missing values**

**Question 6**

- **Find the median value of "Engine Cylinders" column in the dataset.**
- **Next, calculate the most frequent value of the same "Engine Cylinders".**
- **Use the fillna method to fill the missing values in "Engine Cylinders" with the most frequent value from the previous step.**
- **Now, calculate the median value of "Engine Cylinders" once again**


In [10]:
car_price_df['Engine Cylinders'].median()

6.0

In [11]:
car_price_df['Engine Cylinders'].mode()

0    4.0
Name: Engine Cylinders, dtype: float64

In [12]:
#Alternatively to find the most frequent value of the 'Engine Cylinders' columns
car_price_df['Engine Cylinders'].value_counts().idxmax()

4.0

In [13]:
car_price_df['Engine Cylinders'].fillna(4.0, inplace = True)

In [14]:
car_price_df['Engine Cylinders'].median()

6.0

**Answer:**
- **No, the median value has not changed**

**Question 7**
- **Select all the "Lotus" cars from the dataset.**
- **Select only columns "Engine HP", "Engine Cylinders".**
- **Now drop all duplicated rows using drop_duplicates method (you should get a dataframe with 9 rows).**
- **Get the underlying NumPy array. Let's call it X.**
- **Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX**


In [15]:
Lotus_cars_df = car_price_df[car_price_df['Make']=='Lotus']

In [16]:
Lotus_cars_hp_cyln_df = Lotus_cars_df[['Engine HP','Engine Cylinders']]

In [17]:
Lotus_cars_hp_cyln_without_duplicates = Lotus_cars_hp_cyln_df.drop_duplicates()

In [18]:
Lotus_cars_hp_cyln_without_duplicates.shape[0]

9

In [19]:
x = np.array([Lotus_cars_hp_cyln_without_duplicates['Engine HP'].values,Lotus_cars_hp_cyln_without_duplicates['Engine Cylinders'].values])

In [20]:
x

array([[189., 218., 217., 350., 400., 276., 345., 257., 240.],
       [  4.,   4.,   4.,   8.,   6.,   6.,   6.,   4.,   4.]])

In [21]:
x_tranpose = x.T

In [24]:
x_tranpose

array([[189.,   4.],
       [218.,   4.],
       [217.,   4.],
       [350.,   8.],
       [400.,   6.],
       [276.,   6.],
       [345.,   6.],
       [257.,   4.],
       [240.,   4.]])

In [25]:
def vector_vector_multiplication(u, v):
    assert u.shape[1] == v.shape[0]
    
    n = u.shape[0]
    
    result = 0.0
    
    for i in range(n):
        result = result + u[i] *v[i]
    
    return result

In [26]:
def matrix_vector_multiplication(U, v):
    assert U.shape[1] == v.shape[0]
    
    num_rows = U.shape[0]
    
    result = np.zeros(num_rows)
    
    for i in range(num_rows):
        result[i] = vector_vector_multiplication(U[i], v)
    
    return result

In [27]:
def matrix_matrix_multiplication(U, V):
    assert U.shape[1] == V.shape[0]
    
    num_rows = U.shape[0]
    num_cols = V.shape[1]
    
    result = np.zeros((num_rows, num_cols))
    
    for i in range(num_cols):
        vi = V[:, i]
        Uvi = matrix_vector_multiplication(U, vi)
        result[:, i] = Uvi
    
    return result

In [28]:
x_tranpose.shape

(9, 2)

In [29]:
x.shape

(2, 9)

In [30]:
XTX = np.dot(x_tranpose,x)

In [31]:
def invert_matrix_using_gaussian_elimination(U):
    import numpy as np
    n = U.shape[0]
    I = np.identity(n)
    aug_matrix = np.concatenate((U, I), axis = 1)
    for i in range(0,n):
        pivot = aug_matrix[i][i]
        row = aug_matrix[i]
        aug_matrix[i]=row/pivot
        for j in [k for k in range(0,n) if k != i]:
            aug_matrix[j] = aug_matrix[j] - aug_matrix[i] * aug_matrix[j][i]
        inverse_xtx = aug_matrix[:,n:]
        return inverse_xtx
    
    

In [32]:
XTX_inverse = np.linalg.inv(XTX)

LinAlgError: Singular matrix